### ISSUE TITLE
Add styleGuideUrl property to offense when using --display-style-guide and --format=json flags

### ISSUE THREAD

**Issue by vzamanillo** [rubocop-hq/rubocop#7201]
## Is your feature request related to a problem? Please describe.

It is very hard to get the display guide URL from the message, you are forced to parse it (e.g: using regex) which seems very fragile.

## Describe the solution you'd like

Add a property like `styleGuideUrl` (or whatever) when use `--display-style-guide` and --format=json, by this way it should very easy to get and use it (e.g: with editor integrations)

**Comment by gkpacker** [comment]
I'd like to try to implement that

**Comment by vzamanillo** [comment]
I was thinking about the implementation of this feature, I think the better aproach should be adding the Cop reference URLs to the Offense and move most of the logic from MessageAnnotator to another class to handle that URLs and extend it to have a new MessageAnnotator.

**Comment by gkpacker** [comment]
I was studying the code, and i agree that we should add a `style_guide_url` attribute to the offense, so we can get it on `RuboCop::Formatter::JSONFormatter#hash_for_offense` like this:

```rb
      def hash_for_offense(offense)
        offense_hash = {
          severity:  offense.severity.name,
          message:   offense.message,
          cop_name:  offense.cop_name,
          corrected: offense.corrected?,
          location:  hash_for_location(offense)
        }

        offense_hash[:style_guide_url] = offense.style_guide_url if options[:display_style_guide]

        offense_hash
      end
```

What do you think?

**Comment by gkpacker** [comment]
Then, i got an offense on `Offense` parameter list, now that it has more than 5 args, so i was looking for how to use `Config.for_cop` at `Offense.cop_name`, but i don't know if it will work, since its name doesn't have the "department". Does it makes sense?

**Comment by vzamanillo** [comment]
One solution to `Too many parameters` is to use objects as parameters (this implies a lot of codebase changes) or pass the optional parameters as a hash (more appropiated to this case)

```
      def initialize(severity, location, message, cop_name,
                     options = { status: :uncorrected, reference_urls: [] })
        @severity = RuboCop::Cop::Severity.new(severity)
        @location = location
        @message = message.freeze
        @cop_name = cop_name.freeze
        @status = options[:status]
        @reference_urls = options[:reference_urls]
        freeze
      end
```
I wonder if makes sense to check the `display_style_guide` CLI option with JSON format, I think the offense hash should be serialized as the offense is and let the developer choose how to use it, it makes sense with other formatters where the url presence could be obstrusive.

The main advantage of adding the new attribute to the offense is that we let the responsibility to use it (througth MessageAnnotator or what we need) to the formatters, we would not need to annotate the message before adding the offense to the cop anymore

```
      def add_offense(node, location: :expression, message: nil, severity: nil)
        loc = find_location(node, location)

        return if duplicate_location?(loc)

        severity = custom_severity || severity || default_severity

        message ||= message(node)

        # we would not need this anymore
        message = annotate(message)

        status = enabled_line?(loc.line) ? correct(node) : :disabled

        @offenses << Offense.new(severity, loc, message, name, status)
        yield if block_given? && status != :disabled
      end
```

Looks good.

**Comment by gkpacker** [comment]
I'll try to take a look today, but yeah, looks good.
> I wonder if makes sense to check the display_style_guide CLI option with JSON format

Probably not, it was only a primary solution to the problem. I agree that the url property should appear regardless the formatter.

**Comment by gkpacker** [comment]
@vzamanillo Idk if i get it right... Why we wouldn't need to 'annotate' the message anymore? I'm having difficulties on finding where and how i should add the message. I've added the `options` to offense args, but only in the json formatter it's clear where we have to add a key...

**Comment by vzamanillo** [comment]
> @vzamanillo Idk if i get it right... Why we wouldn't need to 'annotate' the message anymore? I'm having difficulties on finding where and how i should add the message. I've added the `options` to offense args, but only in the json formatter it's clear where we have to add a key...

Hey!, we would not need to annotate the message before adding the offense, we would need another way to add the message per formatter (if necessary).
